python/tests/test_conversions.py

```diff
@@ -0,0 +1,288 @@
+            Decimal("1234567891.1234567891234567891234567891"),
+        ),
+        (
+            DecimalType(38, 29),
+            b"\tI\xb0\xf7\x13\xe9\x18[7\xc1x\x0b\x91\xb5$@",
+            Decimal("123456789.12345678912345678912345678912"),
+        ),
+        (
+            DecimalType(38, 30),
+            b"\tI\xb0\xed\x1e\xdf\x80\x03G;\x16\x9b\xf1\x13j\x83",
+            Decimal("12345678.123456789123456789123456789123"),
+        ),
+        (DecimalType(38, 31), b'\tI\xb0\x96+\xac)d(p6)\xea\xc2)"', Decimal("1234567.1234567891234567891234567891234")),
+        (
+            DecimalType(38, 32),
+            b"\tI\xad\xae\xe3h\xe7O\xb5\x14\xbc\xdc+\x95\x9bY",
+            Decimal("123456.12345678912345678912345678912345"),
+        ),
+        (
+            DecimalType(38, 33),
+            b"\tI\x95\x94>5\x93\xde\xb9.\xefS\xb3\xd8\x11\x80",
+            Decimal("12345.123456789123456789123456789123456"),
+        ),
+        (
+            DecimalType(38, 34),
+            b"\tH\xd5\xd7\x90x\xdf\x08\x1a\xf6C\t\x06p\xaf\x07",
+            Decimal("1234.1234567891234567891234567891234567"),
+        ),
+        (DecimalType(38, 35), b"\tCE\x82\x85\xc7Vf$M\x16\x82@f\xd6N", Decimal("123.12345678912345678912345678912345678")),
+        (DecimalType(21, 16), b"\x06\xb1:\xe3\xc4N\x94\xaf\x07", Decimal("12345.1234567891234567")),
+        (DecimalType(22, 17), b"B\xecL\xe5\xab\x11\xce\xd6N", Decimal("12345.12345678912345678")),
+        (DecimalType(23, 18), b"\x02\x9d;\x00\xf8\xae\xb2\x14_\x15", Decimal("12345.123456789123456789")),
+        (DecimalType(24, 19), b"\x1a$N\t\xb6\xd2\xf4\xcb\xb6\xd3", Decimal("12345.1234567891234567891")),
+        (DecimalType(25, 20), b"\x01\x05k\x0ca$=\x8f\xf5$@", Decimal("12345.12345678912345678912")),
+        (DecimalType(26, 21), b"\n6.{\xcbjg\x9f\x93j\x83", Decimal("12345.123456789123456789123")),
+        (DecimalType(27, 22), b'f\x1d\xd0\xd5\xf2(\x0c;\xc2)"', Decimal("12345.1234567891234567891234")),
+        (DecimalType(28, 23), b"\x03\xfd*([u\x90zU\x95\x9bY", Decimal("12345.12345678912345678912345")),
+        (DecimalType(29, 24), b"'\xe3\xa5\x93\x92\x97\xa4\xc7W\xd8\x11\x80", Decimal("12345.123456789123456789123456")),
+        (DecimalType(30, 25), b"\x01\x8e\xe4w\xc3\xb9\xeco\xc9np\xaf\x07", Decimal("12345.1234567891234567891234567")),
+        (DecimalType(31, 26), b"\x0f\x94\xec\xad\xa5C<]\xdePf\xd6N", Decimal("12345.12345678912345678912345678")),
+    ],
+)
+def test_round_trip_conversion_large_decimals(primitive_type, b, result):
+    """Test round trip conversions of calling `conversions.from_bytes` and then `conversions.to_bytes` on the result"""
+    value_from_bytes = conversions.from_bytes(primitive_type, b)
+    assert value_from_bytes == result
+
+    bytes_from_value = conversions.to_bytes(primitive_type, value_from_bytes)
+    assert bytes_from_value == b
+
+
+@pytest.mark.parametrize(
+    "primitive_type, b, result",
+    [
+        (FloatType(), b"\x19\x04\x9e?", 1.2345),
```

Comment:
do these fail if they aren't approximated.  I would think that serializion to and from bytes should be deterministic and not subject to numeric inaccuracies.

Reply:
I kept this consistent with the python legacy client which uses the built-in float and double types and just [tests](https://github.com/apache/iceberg/blob/master/python_legacy/tests/api/test_conversions.py#L48#L51) for approximation. According to the Iceberg spec though the float and double primitive types are supposed to be 32-bit and 64-bit IEEE floating points, so the easiest way I think is to use numpy here which was proposed by @CircArgs at some point.

@dramaticlly this would also require that numpy is used in PR #4262 for `FloatLiteral` and `DoubleLiteral`.

Reply:
Could you elaborate on why is numpy needed here?  

I expect python equality to work, since this is a deterministic round-trip of the data (is it something in the testing framework):
```
>>> 1.2345 == 1.2345
True
```

Reply:
Yes equality works but the imprecision surfaces when converting the float to a c struct and back:
```py
import struct

b = struct.pack("<f", 1.2345)
value = struct.unpack("<f", b)  # 1.2345000505447388
```
and `1.2345 == 1.2345000505447388` returns `False`

Reply:
Interesting, I dug bit further and turned up [this old bug report](https://bugs.python.org/issue4114) which has a couple of solutions to round-trip deterministically (but involves truncation.  I think this is only an issue for Floats and Doubles should round-trip deterministically. I assume the byte values are taken from Java?

Reply:
There will eventually be a write path from python so the byte values stored in the metadata could potentially be from python or Java. But I'd expect today it's almost entirely byte values generated in Java. As the python library matures we'll eventually have some cross-language tests to make sure the behaviors are synced (to_bytes from Java -> from_bytes from python, for example).

That bug report was very helpful, I'm going to try out some of the examples and generally look into this some more. Hopefully there's a simple workaround but I would also be ok with putting a pin in this here and addressing it when we have some of the higher level objects and can measure the impact.

Reply:
Yeah, I don't know that there is a great solution for float, so it is ok to bypass it.  I think double should exactly round-trip.  If it doesn't I need to further refine my understanding.

Reply:
Oh right, yes double does exactly round trip. I moved the double test out of the approximation test function to the equality test function.